### PR TITLE
Build Java cluster on JDK17

### PR DESCRIPTION
**Pull request by mbien**
Changes required to build the java cluster on JDK17 (15 is the latest version NetBeans can be built with right now). I omitted the other clusters to keep the diff smaller.

This won't make the tests work - that would be a larger issue.

note:
this does not include the Gradle version upgrade required to build on JDK17, since 7.3 it is still in the rc phase.
see ~~https://github.com/apache/netbeans/pull/3322~~ https://github.com/apache/netbeans/pull/3326 for the gradle update

### CODE REVIEW

**Comment by lkishalmi** [comment]
I'd close this one in favor of #3322

**Comment by mbien** [comment]
@lkishalmi please reconsider, this PR is not about gradle

**Comment by lkishalmi** [comment]
My bad, have not checked the list of changes.

**Comment by mbien** [comment]
@lkishalmi would you like to review this? This would make the java cluster buildable together with the recently merged gradle PR.

**Comment by mbien** [comment]
@matthiasblaesing @neilcsmith-net would be great to get this in. Any reviewer volunteers? easy review

`ant -Dcluster.config=basic` should finally build after this.

File: harness/apisupport.harness/nbproject/project.properties [hunk 1]
```diff
@@ -16,7 +16,8 @@
 # under the License.
 
 javac.compilerargs=-Xlint -Xlint:-serial
-javac.source=1.6
+javac.source=1.8
+javac.target=1.8
```
Comment (matthiasblaesing) [inline harness/apisupport.harness/nbproject/project.properties:20]: You don't need to set `javac.target` - @JaroslavTulach ensured this in https://github.com/apache/netbeans/commit/4661771. The build system ensures, that `javac.target` is set to the value of `javac.source` is the former is not set.

Comment (mbien) [reply harness/apisupport.harness/nbproject/project.properties:20]: ok. but it doesn't hurt right? does everything use the same build xml which has this functionality? having source and target at different values can cause trouble if its missed somewhere. (ultimately we should aim to use --release in distant future)

i can remove it if you prefer it gone.

Comment (matthiasblaesing) [reply harness/apisupport.harness/nbproject/project.properties:20]: You can see the mechanic in the `build.xml`:

https://github.com/apache/netbeans/blob/61795d673369f587d2ae2f50254b442a407b64c0/ide/css.editor/build.xml#L22-L25

The build looks simple, but truth is, that it just inherits its power from:

-  `nbbuild/templates/projectized.xml` (direct)
-  `nbbuild/templates/common.xml` (transitive)
-  `nbbuild/default.xml`
-  `nbbuild/default-properties.xml`
-  `nbbuild/jdk.xml`
-  `nbbuild/binaries-default-properties.xml`

The `javac.source` and `javac.target` default values are set by the `basic-init` target in `nbbuild/templates/common.xml`. 

From the default target `build` it is reached via:

- `build`
- `netbeans
- `-netbeans`
- `init`
- `basic-init`

For using `--release`, this was also tackled by @JaroslavTulach (I think it is in the same PR) - see here:

https://github.com/apache/netbeans/blob/96720b9d39122e514fe21536a831247e5dc17b10/nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java#L68-L91

This works only for normal module builds, special cases, where the ant `javac` task is called directly, adjustments need to be made there.

NetBeans build is a complex beast, but it some very nice things in it, at the price that its complex.

For the `javac.target`, yes I think it would be good to not add them, else next time, someone will argue, that the two should be adjusted independently from each other.

Comment (mbien) [reply harness/apisupport.harness/nbproject/project.properties:20]: thanks for the info. will remove it.

Comment (mbien) [reply harness/apisupport.harness/nbproject/project.properties]: done, removed `javac.target` from all properties I touched, even if it was there before.

I kept the changes to build xmls since they all had target specified.

Comment (mbien) [reply harness/apisupport.harness/nbproject/project.properties]: it doesn't build. looks like i have to add a few targets again.

Comment (mbien) [reply harness/apisupport.harness/nbproject/project.properties]: ok, it builds again on JDK 17.

please note that this branch is fairly old and i don't want to abuse the CI too much just to rebase and push again. So to get the JDK 17 gradle changes do a
`
git pull --rebase git@github.com:apache/netbeans.git master`

after you checked out this branch.

File: extide/o.apache.tools.ant.module/build.xml [hunk 2]
```diff
@@ -35,7 +35,7 @@
                 <path path="${src-bridge.cp}"/>
             </classpath>
         </depend>
-        <javac srcdir="src-bridge" destdir="build/bridge-classes" deprecation="${build.compiler.deprecation}" debug="${build.compiler.debug}" source="1.6" target="1.6" includeantruntime="false">
+        <javac srcdir="src-bridge" destdir="build/bridge-classes" deprecation="${build.compiler.deprecation}" debug="${build.compiler.debug}" source="1.8" target="1.8" includeantruntime="false">
```
Comment (JaroslavTulach) [inline extide/o.apache.tools.ant.module/build.xml:38]: Consider adding `release="8"` attribute or using `nb-javac` which is supposed to do that automatically.

Comment (mbien) [reply extide/o.apache.tools.ant.module/build.xml:38]: 8 doesn't know `--release`. So i would have to do both, switch to nb-javac and replace both attributes with the release attribute i suppose.

Could we leave this as is until we move to JDK 11+? After that everything will be easier. I am not sure if its worth to put a lot of energy into this, since its only the basic cluster - i expect a lot more changes. It doesn't look like JDK will drop support for source/target any time soon and it does currently allow NB to be built on JDK 17.

Once NB dropped 8 support it would be a fairly simple change. Remove the build support logic Matthias mentioned and switch everything to `javac.release`.

Comment (neilcsmith-net) [reply extide/o.apache.tools.ant.module/build.xml:38]: @mbien the Ant compiler task ignores `release` on JDK 8.

Comment (mbien) [reply extide/o.apache.tools.ant.module/build.xml:38]: @neilcsmith-net interesting! didn't know that. 

plus: "When set and running on JDK 9+ the source and target attributes as well as the bootclasspath will be ignored. Since Ant 1.9.8."

which means we can set all three to the same value? I would be ok with this.

Comment (neilcsmith-net) [reply extide/o.apache.tools.ant.module/build.xml:38]: @mbien yes, as far as I know. Even an argument that only release is necessary, as the only JDK that's going to ignore it will default to source and target being 8 anyway.  But being specific is probably better!

Comment (mbien) [reply extide/o.apache.tools.ant.module/build.xml]: pushed update:
 - `release` requires the new versioning scheme on modern JDKs -> so i have to use 8 instead of 1.8
 - nb-javac throws NPEs when only release is set -> added source and target there too
 - javac now uses only release on all xml files i touched for this PR
 - exception: debugger.jpda.ant can't use release without it causing compiler errors -> kept source/target
 - rebased ontop of latest master, no need to pull gradle changes anymore

Comment (JaroslavTulach) [reply extide/o.apache.tools.ant.module/build.xml]: > set all three to the same value

Looks like Ant's approach is better than Maven's one (fail on JDK8 when release is used) when shielding developers from JDK differences.

> nb-javac throws NPEs when only release is set -> added source and target there too

You can fix the task in `nbbuild/antsrc`...

**Review by JaroslavTulach — APPROVED** [review]
I like usage of `release="8"`!

File: extide/o.apache.tools.ant.module/build.xml (hunk 2 above)
Comment (mbien) [reply extide/o.apache.tools.ant.module/build.xml]: > You can fix the task in nbbuild/antsrc...

yes i do want to take a look later. We shouldn't have a configuration divergence between the two javac tasks if its avoidable.

Comment (mbien) [reply extide/o.apache.tools.ant.module/build.xml]: added 1d773e0

File: webcommon/libs.jstestdriver/build.xml [hunk 3]
```diff
@@ -25,7 +25,7 @@
 
     <target name="ext-compile" depends="build-init">
         <mkdir dir="build/extclasses"/>
-        <javac srcdir="extsrc" destdir="build/extclasses" deprecation="${build.compiler.deprecation}" debug="${build.compiler.debug}" source="1.6" target="1.8">
+        <javac srcdir="extsrc" destdir="build/extclasses" deprecation="${build.compiler.deprecation}" debug="${build.compiler.debug}" release="8">
```
Comment (matthiasblaesing) [inline webcommon/libs.jstestdriver/build.xml:28]: If I read the ant source correctly, this is not correct on JDK 8. You need to set `source`, `target` and `release` to get the desired behavior:

- on JDK 8, `release` will be ignored, as it is invalid for the compiler
- on JDK 9+, `source` and `target` are ignored

Comment (mbien) [reply webcommon/libs.jstestdriver/build.xml:28]: in theory yes. but this would be only a problem if we would compile on jdk 7 or lower too. 8 would have src and target at 8 by default anyway. Everything else can use release already. @neilcsmith-net pointed this out which i didn't consider either initially.

File: nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java [hunk 4]
```diff
@@ -66,21 +66,30 @@ public void setIncludes(String includes) {
 
     @Override
     public void execute() throws BuildException {
-        String tgr = getTarget();
```
Comment (matthiasblaesing) [inline nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java:69]: Why is this necessary?  The original implementation moved target to release - why not keep that behavior?

Comment (mbien) [reply nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java:69]: it still does that if release isn't set, scroll down.

Comment (matthiasblaesing) [reply nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java:69]: No - why setting release? It is not necessary. The original change ensured, that target is transferred to release, unless the bootclasspath is set.

Comment (mbien) [reply nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java:69]: https://github.com/apache/netbeans/pull/3278#discussion_r760538319

Comment (mbien) [reply nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java:69]: + https://github.com/apache/netbeans/pull/3278#discussion_r760775688

Comment (matthiasblaesing) [reply nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java:69]: Not an answer

Comment (mbien) [reply nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java:69]: this change ensures that javac and nb-javac do not diverge configuration wise, as mentioned in the linked conversation.

Comment (mbien) [reply nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java:69]: what is the combination of attributes/JDKs which should be supported but isn't after that change?

**Comment by matthiasblaesing** [comment]
Please check the failing unittests - given the error message:

```
    [junit] INFO: println: /home/travis/build/apache/netbeans/nbbuild/netbeans/harness/common.xml:207: Compile failed; see the compiler error output for details. listener: Hyperlink[file:/home/travis/build/apache/netbeans/nbbuild/netbeans/harness/common.xml:207:-1:-1:-1]
    [junit] Dec 02, 2021 8:37:52 PM org.netbeans.modules.apisupport.project.InputOutputProviderImpl$OW write
    [junit] INFO: BUILD FAILED (total time: 1 second)
```

This might indeed a problem.

**Comment by mbien** [comment]
i see many of those messages on a build which doesn't fail. E.g taken from latest master build:
https://app.travis-ci.com/github/apache/netbeans/jobs/550425157#L9197-L9199

its difficult to figure out what is failing in a build log full with "build failed" unfortunately

**Comment by matthiasblaesing** [comment]
Thank you - travis and github actions are happy now. Would you please squash the commits into a single one?

**Comment by mbien** [comment]
done. Maybe @JaroslavTulach would want to take another look since he approved before I changed the ant task.

File: nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java [hunk 5]
```diff
@@ -66,21 +66,30 @@ public void setIncludes(String includes) {
...
-        String src = getSource();
-        if (src.matches("\\d+")) {
-            src = "1." + src;
-        }
-        if (!JavaEnvUtils.isAtLeastJavaVersion(src)) {
-            log("Cannot handle -source " + src + " from this VM; forking " + maybeFork, Project.MSG_WARN);
-            super.setFork(true);
-            super.setExecutable(maybeFork);
+        String release = getRelease();
+        if (release == null || release.isEmpty()) {
```
Comment (JaroslavTulach) [inline nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java:70]: Setting the `release` only when it is not set yet seems OK to me.

**Review by JaroslavTulach — APPROVED** [review]

File: nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java (hunk 5 above)
Comment (mbien) [reply nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java:70]: thanks for taking a look jaroslav!

**Comment by matthiasblaesing** [comment]
Ok - travis is happy, lets get this in.

**Comment by JaroslavTulach** [comment]
FYI: [Jenkins VSCode build #739](https://ci-builds.apache.org/job/Netbeans/job/netbeans-vscode/739/) failed and continues to fail. Coincidence?

**Comment by matthiasblaesing** [comment]
#739 fails, but #742 was green. Given, that the test where the timeout occurs is marked @RandomlyFails, a broken tests sounds like a reasonable explanation.

**Comment by JaroslavTulach** [comment]
A coincidence, then. Sorry for bothering.
